### PR TITLE
[BC Break] Migrate from redux-form to react-final-form

### PR DESCRIPTION
**Pull request by djhi**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -40,7 +40,6 @@
     "devDependencies": {
         "@types/jest": "^24.0.13",
         "@types/react": "^16.8.0",
-        "@types/react-redux": "^7.1.0",
```
Comment (fzaninotto) [inline package.json]: why did you remove that?

File: packages/ra-core/package.json [hunk 2]
```diff
@@ -30,11 +30,9 @@
         "@redux-saga/testing-utils": "^1.0.2",
         "@types/history": "^4.7.2",
         "@types/node-polyglot": "^0.4.31",
-        "@types/react-redux": "^7.1.1",
```
Comment (fzaninotto) [inline packages/ra-core/package.json]: why

File: examples/simple/src/posts/PostCreate.js [hunk 3]
```diff
@@ -21,20 +21,20 @@ import {
...
 
+    const form = useForm();
+    const values = form.getState().values;
     const handleClick = useCallback(
-        () =>
-            handleSubmit(values => {
-                dispatch(saveWithNote(values, basePath, redirect));
-            }),
-        [basePath, dispatch, handleSubmit, redirect]
+        () => () => dispatch(saveWithNote(values, basePath, redirect)),
```
[the remark this answers is not in the document]
Comment (djhi) [reply examples/simple/src/posts/PostCreate.js]: I'll have to check, I think it comes from how redux-form worked before

File: package.json (hunk 1 above)
Comment (djhi) [reply package.json]: It was a very hot day and my melted brain mistaken it for redux-form types...

File: examples/demo/src/reviews/RejectButton.js [hunk 4]
```diff
@@ -1,27 +1,46 @@
...
-import { formValueSelector } from 'redux-form';
 import Button from '@material-ui/core/Button';
 import ThumbDown from '@material-ui/icons/ThumbDown';
-import { useTranslate } from 'react-admin';
-import { reviewReject as reviewRejectAction } from './reviewActions';
+import { useTranslate, useUpdate } from 'react-admin';
 
+const options = {
+    undoable: true,
+    onSuccess: {
```
Comment (fzaninotto) [inline examples/demo/src/reviews/RejectButton.js]: use function side effects

File: examples/simple/src/posts/PostCreate.js [hunk 5]
```diff
@@ -18,24 +16,36 @@ import {
...
+            {
+                data: { ...formState.values, average_note: 10 },
+            },
+            {
+                onSuccess: ({ data: newRecord }) => {
+                    redirectTo(redirect, basePath, newRecord.id, newRecord);
+                },
+            }
+        );
+    }, [create, form, redirect, redirectTo, basePath]);
```
Comment (fzaninotto) [inline examples/simple/src/posts/PostCreate.js]: I'm afraid the `form` object never changes (only its properties), and `useCallback` does identity check. Can't this cause a bug? shouldn't you track the values instead?

File: packages/ra-core/src/controller/useEditController.spec.tsx [hunk 6]
```diff
@@ -47,17 +47,6 @@ describe('useEditController', () => {
         expect(getByText('hello')).toBeDefined();
     });
 
-    it('should reset the redux form', () => {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/useEditController.spec.tsx]: the test used to test the implementation, but the feature 'should reset the ~~redux~~ form' still needs to be tested, don't you think?

File: packages/ra-core/src/controller/useEditController.spec.tsx [hunk 7]
```diff
@@ -69,7 +58,7 @@ describe('useEditController', () => {
             </EditController>
         );
         act(() => saveCallback({ foo: 'bar' }));
-        const crudUpdateAction = dispatch.mock.calls[2][0];
+        const crudUpdateAction = dispatch.mock.calls[1][0];
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/useEditController.spec.tsx]: instead of relying on the dispatch mock call index, I tend to do a `find` using the action type to make the test future proof. Not necessary for this PR though.

File: packages/ra-core/src/controller/useEditController.ts [hunk 8]
```diff
@@ -74,10 +70,6 @@ const useEditController = (props: EditProps): EditControllerProps => {
         },
     });
 
-    useEffect(() => {
-        dispatch(resetForm(REDUX_FORM_NAME));
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/useEditController.ts]: this was somehow necessary for the create to edit redirection...

File: packages/ra-core/src/controller/useListController.spec.tsx [hunk 9]
```diff
@@ -88,37 +88,6 @@ describe('useListController', () => {
             });
         });
 
-        it('should not call filtering function if filters are unchanged', () => {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/useListController.spec.tsx]: why did you remove that test?

File: packages/ra-core/src/createAdminStore.ts [hunk 10]
```diff
@@ -61,11 +60,7 @@ export default ({
         resettableAppReducer,
         initialState,
         compose(
-            applyMiddleware(
-                sagaMiddleware,
-                formMiddleware,
-                routerMiddleware(history)
-            ),
+            applyMiddleware(sagaMiddleware, routerMiddleware(history)),
```
Comment (fzaninotto) [inline packages/ra-core/src/createAdminStore.ts]: this is a breaking change for custom apps, it should be documented in the custom apps and the upgrade guide

File: packages/ra-core/src/form/FormDataConsumer.tsx [hunk 11]
```diff
@@ -119,21 +119,13 @@ export const FormDataConsumerView: SFC<Props> = ({
...
-
-const ConnectedFormDataConsumerView = connect(mapStateToProps)(
-    FormDataConsumerView
-);
-
 const FormDataConsumer = (props: ConnectedProps) => (
-    <FormName>
-        {({ form }) => <ConnectedFormDataConsumerView form={form} {...props} />}
-    </FormName>
+    <FormSpy
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormDataConsumer.tsx]: I'd prefer `useFormState` to avoid render props

File: examples/simple/src/posts/PostCreate.js [hunk 12]
```diff
@@ -18,24 +16,36 @@ import {
...
+    const { basePath, redirect } = props;
 
-    const handleClick = useCallback(
-        () =>
-            handleSubmit(values => {
-                dispatch(saveWithNote(values, basePath, redirect));
-            }),
-        [basePath, dispatch, handleSubmit, redirect]
-    );
+    const form = useForm();
```
Comment (fzaninotto) [inline examples/simple/src/posts/PostCreate.js]: useFormState?

File: packages/ra-core/src/form/FormField.tsx [hunk 13]
```diff
@@ -15,32 +13,34 @@ export const isRequired = validate => {
...
 }
 
-export const FormFieldView: SFC<Props> = ({ input, ...props }) =>
-    input ? ( // An ancestor is already decorated by Field
+export const FormField: SFC<Props> = ({ input, validate, ...props }) => {
+    const sanitizedValidate = Array.isArray(validate)
+        ? composeValidators(validate)
+        : validate;
+
+    return input ? ( // An ancestor is already decorated by Field
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormField.tsx]: contrary to redux-form, I don't think react-final-form injects an `input` prop, so you'll need another trick to avoid double decoration

File: packages/ra-core/src/form/formMiddleware.ts [hunk 14]
```diff
@@ -1,45 +0,0 @@
-import { LOCATION_CHANGE } from 'connected-react-router';
-import { destroy } from 'redux-form';
-import isEqual from 'lodash/isEqual';
-
-import { resetForm } from '../actions/formActions';
-import { REDUX_FORM_NAME } from './constants';
-
-/**
- * This middleware ensure that whenever a location change happen, we get the
```
Comment (fzaninotto) [inline packages/ra-core/src/form/formMiddleware.ts]: Do we have tests that prove that this still works?

File: packages/ra-core/src/form/validate.ts [hunk 15]
```diff
@@ -60,6 +60,20 @@ type Memoize = <T extends (...args: any[]) => any>(
 const memoize: Memoize = (fn: any) =>
     lodashMemoize(fn, (...args) => JSON.stringify(args));
 
+// Compose multiple validators into a single one for use with final-form
+export const composeValidators = (...validators) => (value, values, meta) => {
```
Comment (fzaninotto) [inline packages/ra-core/src/form/validate.ts]: I'd love a unit test for that one

File: packages/ra-core/src/reducer/index.ts [hunk 16]
```diff
@@ -13,7 +12,6 @@ export default (customReducers, locale, messages, history) =>
     combineReducers({
         admin,
         i18n: i18nReducer(locale, messages),
-        form: formReducer,
```
Comment (fzaninotto) [inline packages/ra-core/src/reducer/index.ts]: this should be documented in custom apps and upgrade

File: packages/ra-core/src/sideEffect/redirection.spec.ts [hunk 17]
```diff
@@ -34,12 +33,4 @@ describe('redirection saga', () => {
         const generator = handleRedirection(action);
         expect(generator.next().value).toEqual(put(push('/posts/123')));
     });
-
-    it('should yield a form refresh if redirectTo is falsy', () => {
```
Comment (fzaninotto) [inline packages/ra-core/src/sideEffect/redirection.spec.ts]: How do we implement the "save and add" then?

File: packages/ra-core/src/sideEffect/redirection.ts [hunk 18]
```diff
@@ -41,27 +40,28 @@ export function* handleRedirection({
...
-                          ? payload.data
-                          : requestPayload && requestPayload.data
-                          ? requestPayload.data
-                          : null
-                  )
-              )
-          )
-        : yield put(reset('record-form')); // explicit no redirection, reset the form
+    if (!redirectTo) {
+        return;
```
Comment (fzaninotto) [inline packages/ra-core/src/sideEffect/redirection.ts]: Why is that different from the `useRedirect` hook logic?

File: packages/ra-core/src/sideEffect/useRedirect.ts [hunk 19]
```diff
@@ -37,12 +37,14 @@ const useRedirect = () => {
...
             data?: Partial<Record>
-        ) =>
-            redirectTo
-                ? dispatch(
-                      push(resolveRedirectTo(redirectTo, basePath, id, data))
-                  )
-                : dispatch(reset('record-form')), // explicit no redirection, reset the form
+        ) => {
+            if (!redirectTo) {
+                dispatch(refreshView());
```
Comment (fzaninotto) [inline packages/ra-core/src/sideEffect/useRedirect.ts]: I'm surprised that a simple refresh is enough to empty a create form. Do you confirm it works?

File: packages/ra-core/src/util/TestContext.spec.tsx [hunk 20]
```diff
@@ -70,48 +69,48 @@ describe('TestContext.js', () => {
...
-                    {({ store }) => {
-                        testStore = store;
-                        return <span>foo</span>;
-                    }}
-                </TestContext>
-            );
-            const initialstate = testStore.getState();
-            initialstate.router.location.key = ''; // react-router initializes the state with a random key
-            expect(initialstate).toEqual(primedStore);
+    // describe('enableReducers options', () => {
```
Comment (fzaninotto) [inline packages/ra-core/src/util/TestContext.spec.tsx]: commented out code :cold_sweat:

File: packages/ra-ui-materialui/src/auth/LoginForm.tsx [hunk 21]
```diff
@@ -29,33 +18,24 @@ interface FormData {
...
+    button: {
+        width: '100%',
+    },
+    icon: {
+        marginRight: theme.spacing(1),
+    },
+}));
 
-// see http://redux-form.com/6.4.3/examples/material-ui/
 const renderInput = ({
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/auth/LoginForm.tsx]: this is a Component, it should start with a capital. And the name doesn't feel right.

File: packages/ra-ui-materialui/src/form/FormTab.js [hunk 22]
```diff
@@ -20,7 +20,7 @@ const hiddenStyle = { display: 'none' };
 
 class FormTab extends Component {
     renderHeader = ({ className, label, icon, value, translate, ...rest }) => {
-        const to = { pathname: value, search: 'skipFormReset' }; // FIXME use location state when https://github.com/supasate/connected-react-router/issues/301 is fixed
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/form/FormTab.js]: The issue isn't fixed.
Oh wait, is that because you removed the saga watching on the location change to reset the form?

File: packages/ra-ui-materialui/src/form/SimpleForm.js [hunk 23]
```diff
@@ -49,72 +56,89 @@ const sanitizeRestProps = ({
...
     validate,
+    validating,
     _reduxForm,
     ...props
 }) => props;
 
-export class SimpleForm extends Component {
-    handleSubmitWithRedirect = (redirect = this.props.redirect) =>
-        this.props.handleSubmit(values => this.props.save(values, redirect));
+export const SimpleForm = ({
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/form/SimpleForm.js]: I was confused when I saw that this SimpleForm doesn't include a Form. I think you should inverse the naming of the two components in this file, and export SimpleForm which would call a SimpleFormView. Also, please inverse the two components so that the reader goes from the top down.

File: packages/ra-ui-materialui/src/form/SimpleForm.js [hunk 24]
```diff
@@ -49,72 +56,89 @@ const sanitizeRestProps = ({
...
+    version,
+    handleSubmit,
+    ...rest
+}) => {
+    useEffect(() => {
+        if (!record) {
+            return;
+        }
+
+        form.batch(() => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/form/SimpleForm.js]: Could you comment that block? It's cryptic.

File: examples/demo/src/reviews/reviewSaga.js [hunk 25]
```diff
@@ -1,15 +0,0 @@
-import { all, takeEvery } from 'redux-saga/effects';
```
[the remark this answers is not in the document]
Comment (djhi) [reply examples/demo/src/reviews/reviewSaga.js]: Yes but we don't even use the custom review actions

File: packages/ra-ui-materialui/src/form/SimpleForm.js [hunk 26]
```diff
@@ -138,18 +162,48 @@ SimpleForm.defaultProps = {
...
-    translate, // Must be before reduxForm so that it can be used in validation
-    reduxForm({
-        destroyOnUnmount: false,
-        enableReinitialize: true,
-        keepDirtyOnReinitialize: true,
-    })
-);
+const EnhancedSimpleForm = ({ initialValues, ...props }) => {
+    let redirect;
+    // We don't use state here for two reasons:
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/form/SimpleForm.js]: why don't you use a ref?

File: packages/ra-ui-materialui/src/form/TabbedForm.js [hunk 27]
```diff
@@ -72,121 +73,122 @@ export const getTabFullPath = (tab, index, baseUrl) =>
         tab.props.path ? `/${tab.props.path}` : index > 0 ? `/${index}` : ''
     }`;
 
-export class TabbedForm extends Component {
-    handleSubmitWithRedirect = (redirect = this.props.redirect) =>
-        this.props.handleSubmit(values => this.props.save(values, redirect));
+export const TabbedForm = ({
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/form/TabbedForm.js]: Same remark concerning the 2 components as in Simpleform

File: packages/ra-ui-materialui/src/form/TabbedForm.js [hunk 28]
```diff
@@ -72,121 +73,122 @@ export const getTabFullPath = (tab, index, baseUrl) =>
...
+    value,
+    version,
+    ...rest
+}) => {
+    useEffect(() => {
+        if (!record) {
+            return;
+        }
+
+        form.batch(() => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/form/TabbedForm.js]: this is magic

File: packages/ra-ui-materialui/src/form/TabbedForm.js [hunk 29]
```diff
@@ -72,121 +73,122 @@ export const getTabFullPath = (tab, index, baseUrl) =>
...
-                                                   * Force redraw when the tab becomes active
-                                                   *
-                                                   * This is because the fields, decorated by redux-form and connect,
-                                                   * aren't redrawn by default when the tab becomes active.
-                                                   * Unfortunately, some material-ui fields (like multiline TextField)
-                                                   * compute their size based on the scrollHeight of a dummy DOM element,
-                                                   * and scrollHeight is 0 in a hidden div. So they must be redrawn
-                                                   * once the tab becomes active.
-                                                   *
-                                                   * @ref https://github.com/marmelab/react-admin/issues/1956
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/form/TabbedForm.js]: Can you confirm this issue doesn't come back?

File: packages/ra-ui-materialui/src/form/TabbedForm.js [hunk 30]
```diff
@@ -255,36 +239,50 @@ export const findTabsWithErrors = (
...
-            (acc, child) => [
-                ...acc,
-                ...(isValidElement(child)
-                    ? Children.toArray(child.props.children)
-                    : []),
-            ],
-            []
-        );
+const EnhancedTabbedForm = ({ initialValues, ...props }) => {
+    let redirect;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/form/TabbedForm.js]: useRef?

File: packages/ra-ui-materialui/src/input/ArrayInput.spec.js [hunk 31]
```diff
@@ -1,118 +1,118 @@
...
-                    <SimpleFormIterator>
-                        <NumberInput source="id" />
-                        <TextInput source="foo" />
-                    </SimpleFormIterator>
-                </ArrayInput>
-            </form>
-        );
-        const DummyFormRF = reduxForm({ form: 'record-form' })(DummyForm);
+    // FIXME
+    it.skip('should clone each input once per value in the array', () => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput.spec.js]: skip detected

File: packages/ra-ui-materialui/src/list/FilterForm.js [hunk 32]
```diff
@@ -127,36 +131,53 @@ export const mergeInitialValuesWithDefaultValues = ({
...
-        enableReinitialize: true,
-        destroyOnUnmount: false, // do not destroy to preserve state across navigation
-        onChange: (values, dispatch, props) =>
-            props && props.setFilters(values),
-    })
-);
-
-export default enhance(FilterForm);
+const FormSpySubscription = { values: true, pristine: true };
+const EnhancedFilterForm = props => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/FilterForm.js]: please inverse the 2 components to make the file more readable

File: packages/ra-ui-materialui/src/list/FilterForm.js [hunk 33]
```diff
@@ -127,36 +131,53 @@ export const mergeInitialValuesWithDefaultValues = ({
...
-        form: 'filterForm',
-        enableReinitialize: true,
-        destroyOnUnmount: false, // do not destroy to preserve state across navigation
-        onChange: (values, dispatch, props) =>
-            props && props.setFilters(values),
-    })
-);
-
-export default enhance(FilterForm);
+const FormSpySubscription = { values: true, pristine: true };
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/FilterForm.js]: What's that? A comment would be welcome

File: packages/ra-ui-materialui/src/form/TabbedForm.js (hunk 29 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/form/TabbedForm.js]: I tried it yes

**Comment by djhi** [comment]
> I wonder if we're not missing one part: the rewrite of all Input fields to remove their addField HOC. What do you think?

react-final-form still provide a `Field` component so we still need the HOC, right ?

File: packages/ra-ui-materialui/src/form/TabbedForm.js (hunk 27 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/form/TabbedForm.js]: I don't get this comment

File: packages/ra-core/src/controller/useEditController.spec.tsx (hunk 6 above)
Comment (djhi) [reply packages/ra-core/src/controller/useEditController.spec.tsx]: We have e2e tests for it

File: packages/ra-core/src/controller/useEditController.ts (hunk 8 above)
Comment (djhi) [reply packages/ra-core/src/controller/useEditController.ts]: It's not necessary anymore

File: packages/ra-core/src/controller/useListController.spec.tsx (hunk 9 above)
Comment (djhi) [reply packages/ra-core/src/controller/useListController.spec.tsx]: Because I removed the equality check in the setFilters function. Could not make it work otherwise

File: packages/ra-core/src/form/FormField.tsx (hunk 13 above)
Comment (djhi) [reply packages/ra-core/src/form/FormField.tsx]: Nope, it does: https://github.com/final-form/react-final-form#fieldrenderprops

File: packages/ra-core/src/form/formMiddleware.ts (hunk 14 above)
Comment (djhi) [reply packages/ra-core/src/form/formMiddleware.ts]: Yes, there are e2e tests for it

File: packages/ra-core/src/sideEffect/redirection.spec.ts (hunk 17 above)
Comment (djhi) [reply packages/ra-core/src/sideEffect/redirection.spec.ts]: It works on the simple example ;)

File: packages/ra-core/src/sideEffect/useRedirect.ts (hunk 19 above)
Comment (djhi) [reply packages/ra-core/src/sideEffect/useRedirect.ts]: 1. What is a simple refresh ?
2. Yes and we have e2e tests for all redirection cases (see `PostCreate` which redirect to `edit`, `show`, `create`)

File: packages/ra-core/src/form/FormDataConsumer.tsx [hunk 34]
```diff
@@ -119,21 +117,10 @@ export const FormDataConsumerView: SFC<Props> = ({
     return ret === undefined ? null : ret;
 };
 
-const mapStateToProps = (
-    state: ReduxState,
-    { form, record }: ConnectedProps
-) => ({
-    formData: getFormValues(form)(state) || record,
-});
+const FormDataConsumer = (props: ConnectedProps) => {
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormDataConsumer.tsx]: seeing how little this component brings, I suggest we deprecate it by removing it of the documentation.

File: packages/ra-core/src/form/validate.spec.ts [hunk 35]
```diff
@@ -20,6 +21,45 @@ describe('Validators', () => {
                 .map(error => (error && error.message ? error.message : error)),
             Array(...Array(inputs.length)).map(() => message)
         );
+
+    describe('composeValidators', () => {
+        it('Correctly compose validators passed as an array', () => {
```
Comment (fzaninotto) [inline packages/ra-core/src/form/validate.spec.ts]:
```suggestion
        it('Correctly composes validators passed as an array', () => {
```

File: packages/ra-core/src/form/validate.spec.ts [hunk 36]
```diff
@@ -20,6 +21,45 @@ describe('Validators', () => {
...
+                'ra.validation.minLength'
+            );
+            test(
+                composeValidators([required(), minLength(5)]),
+                ['abcde'],
+                undefined
+            );
+        });
+
+        it('Correctly compose validators passed as many arguments', () => {
```
Comment (fzaninotto) [inline packages/ra-core/src/form/validate.spec.ts]:
```suggestion
        it('Correctly composes validators passed as many arguments', () => {
```

File: packages/ra-ui-materialui/src/auth/LoginForm.tsx [hunk 37]
```diff
@@ -29,33 +18,24 @@ interface FormData {
...
+    },
+}));
 
-// see http://redux-form.com/6.4.3/examples/material-ui/
-const renderInput = ({
-    meta: { touched, error } = { touched: false, error: '' }, // eslint-disable-line react/prop-types
-    input: { ...inputProps }, // eslint-disable-line react/prop-types
+const Input = ({
+    meta: { touched, error }, // eslint-disable-line react/prop-types
+    input: inputProps, // eslint-disable-line react/prop-types
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/auth/LoginForm.tsx]: I think `input` no longer exists with react-final-form

File: packages/ra-ui-materialui/src/list/FilterForm.js (hunk 32 above)
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/list/FilterForm.js]: bump

File: packages/ra-core/src/controller/useListController.spec.tsx (hunk 9 above)
Comment (fzaninotto) [reply packages/ra-core/src/controller/useListController.spec.tsx]: What's the impact?

File: packages/ra-ui-materialui/src/auth/LoginForm.tsx (hunk 37 above)
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/auth/LoginForm.tsx]: sorry, disregard my comment.

File: packages/ra-core/src/controller/useListController.spec.tsx (hunk 9 above)
Comment (djhi) [reply packages/ra-core/src/controller/useListController.spec.tsx]: I have not seen any but as we already know, a global performance check will be needed for the 3.0

File: packages/ra-ui-materialui/src/list/FilterForm.js (hunk 32 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/list/FilterForm.js]: This is weird, I inverted them already

File: packages/ra-core/src/form/FormDataConsumer.tsx (hunk 34 above)
Comment (djhi) [reply packages/ra-core/src/form/FormDataConsumer.tsx]: As discussed, we'll see that in another PR

File: packages/ra-core/src/reducer/index.ts (hunk 16 above)
Comment (djhi) [reply packages/ra-core/src/reducer/index.ts]: Done

File: packages/ra-core/src/createAdminStore.ts (hunk 10 above)
Comment (djhi) [reply packages/ra-core/src/createAdminStore.ts]: Done
